workflow/validate/validate.go

```diff
@@ -163,10 +198,27 @@ func (ctx *wfValidationCtx) validateTemplate(tmpl *wfv1.Template, args wfv1.Argu
 	return nil
 }
 
+// validateTemplateRef validates template reference.
+func validateTemplateRef(tmpl *wfv1.Template) error {
+	if tmpl.GetType() != wfv1.TemplateTypeUnknown {
+		return errors.New(errors.CodeBadRequest, "template ref can not be used with template type.")
+	}
+	if tmpl.Template != "" {
+		return errors.Errorf(errors.CodeBadRequest, "template cannot be specified with templateRef")
+	}
+	if tmpl.TemplateRef.Name == "" {
+		return errors.Errorf(errors.CodeBadRequest, "resource name is required")
+	}
+	if tmpl.TemplateRef.Template == "" {
+		return errors.Errorf(errors.CodeBadRequest, "template name is required")
+	}
+	return nil
+}
+
 // validateTemplateType validates that only one template type is defined
 func validateTemplateType(tmpl *wfv1.Template) error {
 	numTypes := 0
-	for _, tmplType := range []interface{}{tmpl.Container, tmpl.Steps, tmpl.Script, tmpl.Resource, tmpl.DAG, tmpl.Suspend} {
+	for _, tmplType := range []interface{}{tmpl.TemplateRef, tmpl.Container, tmpl.Steps, tmpl.Script, tmpl.Resource, tmpl.DAG, tmpl.Suspend} {
```

Comment:
The error message for an unspecified template type should include `templateRef`, for now it only mentions `container, steps, script, resource, dag, suspend`.